### PR TITLE
WIP: Consolidate patches and add py3k patches

### PR DESCRIPTION
**Pull request by jakirkham**
After some hacking around on the source and using swig, it seems we may have gotten Python 3.x support for this library. It needs some better testing. However, my testing locally suggests this may work. Let's see if this builds ok on CI at a minimum.

This includes some patches from Arch Linux ( https://aur.archlinux.org/cgit/aur.git/tree/?h=python-spams-svn ). However, they needed some tweaks. It turned out the easiest path forward was to change the SWIG files and regenerate the wrapper code. The result of this was transformed into patches.

### CODE REVIEW

File: recipe/meta.yaml [hunk 1]
```diff
@@ -14,21 +14,21 @@ source:
...
-    - spams_wrap.cpp.patch
-    - spams_wrap-windows.cpp.patch
+    # Support Python 3.x.
+    - spams_py3k_np10.diff
 
 build:
-  skip: true  # [win or py3k]
-  number: 200
+  skip: true  # [win]
+  number: 201
```
Comment (jakirkham) [inline recipe/meta.yaml]: Need to bump this.

**Comment by samuelstjean** [comment]
I was looking for exactly this, if you need help I also re-hosted myself a patched version using the files from arch + latest swig (which works on my arch based laptop for what it is worth) over there https://github.com/samuelstjean/spams-python/releases/download/0.1/spams-python-v2.5-python3.zip

Just in case it might help, where is your build trouble happening? The version I linked here works fine on linux, unfortunately can't say for windows and mac. I actually doubt it would work on windows as-is, the base setup.py seems a bit hackish for this one as it pulls stuff from R.

**Comment by jakirkham** [comment]
So, I actually mirrored what is present from the SVN repo to this GitHub repo ( https://github.com/spams-org/spams-devel ). Was talking to one of the developers about it. Sounds like work on the original code base has stopped at present. So I'm trying to decide if I should fork it and do a release with these patches. Would welcome other people like yourself that are interested in seeing it keep going.

This patch is basically generated from this [branch on my fork]( https://github.com/jakirkham/spams-devel/tree/python_3_5_patches ) followed by a run of SWIG. The patch is very similar to the Arch Linux one IIRC. Ideally I would want to start generating this directly from source. So have some issues to sort out with SWIG if you are interested in helping. Unfortunately, it has been awhile since I looked. So would need to refresh myself as well as to why this was an issue.

**Comment by samuelstjean** [comment]
Ah well, if you are familiar with the patching and build system of arch, it is just a fancier bash script. I generated this file by cloning the latest svn, using the latest swig, applying the python3 arch patch and zipping the whole thing (all part are automated, execept the last one).

Since the setup.py is fairly untouched, it does not produce valid wheels/sdist when run, but it does pip install with distutils though. That's why I think it would require a bit of work to run on windows, and then I found this repo which was trying to do just that. 

I did not know they themselves had a github repo for it though, my fork predates that a bit for easy pip install as a dep of other projects, but this python3 file is form the latest svn even though the other download are just a rehost. Well if you ever wanna get this up and running again I can probably pitch in also.

**Comment by zym1010** [comment]
may I ask what is the status of this Py3k patching project? Thanks.

**Comment by jakirkham** [comment]
In conda-forge, after a few discussions, it seems people don't like patching things to add Python 3 support. It should be possible with SWIG to avoid this dispute entirely. However I have lacked the time to pursue it. If someone wants to try a PR along those lines, I would be happy to explore it with them.

**Comment by samuelstjean** [comment]
They also have a new 2.6 release with python 3 patches already applied
internally (and the git version has python 3.6/numpy 1.12 support).
It could be useful to try to upstream the work already done, their patches
are pretty much the ones from the archlinux AUR + small other stuff.

2017-04-06 15:47 GMT+02:00 jakirkham <notifications@github.com>:

**Comment by jakirkham** [comment]
I had emailed them a couple years back with some patches and explanations to which they later responded. It was unclear at the time whether there would be further development or not (not uncommon with academic projects like this one). So it's good to hear there was and they addressed a few things. Maybe they even used the same patches. If they have a new release that fixes things, it makes more sense to try that then.

**Comment by samuelstjean** [comment]
They also have a public gitlab over here
https://gitlab.inria.fr/thoth/spams-devel/tree/master

I sent them some patches to include, so that's why I guessed they used them
(partly they mentioned). Anyway, you also had some stuff about making it
build on windows, so maybe you could try sending them that to ease
maintenance in here.

This person is now in charge of the python version :
https://gitlab.inria.fr/gdurif
You can find his email easily over there gitlab, which I won't post here
because bots.

2017-04-06 16:12 GMT+02:00 jakirkham <notifications@github.com>:

**Comment by zym1010** [comment]
is it possible that we simply base the py3k version of spams on their official py3k branch, since they have done all the hard work for it?

**Comment by jakirkham** [comment]
I thought they already have a release for it. Is that correct @samuelstjean?

**Comment by samuelstjean** [comment]
Yes, they have a python and python 3 folder, I find it personally a bit hard to follow though as the script calls symlink all over the folders and they have a bunch of readmes everywhere.

This is the main patch file https://gitlab.inria.fr/thoth/spams-devel/blob/master/swig/python/python.patch

Current release spams 2.6 is split between py2 and py3, but they want to have a single one for the next one (presumably called spams 2.7). The git version also works for numpy 1.12 (and 2.6 does not if memory serves), some fortran ordering stuff was broken due to numpy changes.

**Comment by samuelstjean** [comment]
Ah yes on their official page
http://spams-devel.gforge.inria.fr/downloads.html

Works for numpy < 1.12 and no windows build though :/ But hey at least it
is still alive instead of people like us patching left and right and making
non compatible version (I also have my own custom patches and fork which
nobody except me (should) use).

Hopefully we can pool together stuff and make it work upstream for good.

2017-04-06 16:51 GMT+02:00 jakirkham <notifications@github.com>:

**Comment by jakirkham** [comment]
Alright, closing this out. Working on releasing 2.6 in PR ( https://github.com/conda-forge/python-spams-feedstock/pull/11 ). Would appreciate your feedback over there.

**Comment by zym1010** [comment]
@samuelstjean 

> Works for numpy < 1.12 and no windows build though

where do you find that it can't work with numpy 1.12?

**Comment by jakirkham** [comment]
I thinks @samuelstjean is referring to this [patch]( https://github.com/conda-forge/python-spams-feedstock/blob/a7a809d2542d101b92e2467dd0de7e93231f1733/recipe/spams_wrap.cpp.patch ) that we need to work with NumPy. Based on my notes, this was necessary to even build with NumPy 1.10 and is still needed for NumPy 1.11 and 1.12. Have updated the patch for 2.6 (as things moved around) and it seems to work with 1.11 and 1.12. Would note that this patch would not even be necessary if we used `swig` to build this package directly from source.

**Comment by samuelstjean** [comment]
That's possible regarding the patch, my comment referred to the fact that the original version (currently released) had not this, and it was on the deprecation route since numpy 1.10. Starting with 1.12, the way they check strides changed and it would build, but crash midway when accessing any function doing linear algebra (i.e. everything).

The guy from spams fixed it in the git version, which is not released yet, but it is fully possible that this patch fixes it, I have to admit I never tried this particular build here. The upstream git fixes the ordering, but some tests won't run due to something else breaking with numpy 1.12 changes (as opposed to everything without the patch).

**Comment by jakirkham** [comment]
Well, if you figure out what the change was, I'd be happy to tack on a patch in the 2.6 release. ( https://github.com/conda-forge/python-spams-feedstock/pull/11 )

**Comment by jakirkham** [comment]
So the only thing I'm seeing w.r.t. NumPy striding issues is this [commit]( https://gitlab.inria.fr/thoth/spams-devel/commit/37ffd87e6735b61087e5e9d65c21ed7d20a8e344 ), which we are definitely including in our [patch]( https://github.com/conda-forge/python-spams-feedstock/blob/a7a809d2542d101b92e2467dd0de7e93231f1733/recipe/spams_wrap.cpp.patch#L14 ). Please let me know if this what you were referring to or if it was something else, @samuelstjean.

**Comment by samuelstjean** [comment]
Seems to be it, would not be surprised if it is your patch in the first place (had I realized that I would not have struggled on my side to make it work...), I told them to go fetch some of your stuff here as needed indeed. But yeah that would have been the main offender I'd guess.

**Comment by jakirkham** [comment]
Cool. Funny that you should mention collaborating with them. I just sent an email to Ghislain with the address I could find (an Inria one?). Anyways maybe we can start discussing migrating patches over there way. Can cc anyone who has interest on the email thread.

**Comment by samuelstjean** [comment]
Yep that's the guy, you can add me if you want, i've been testing that the stuff is numpy 1.12 compliant already, so he should remember me. Apparently they picked it back up for development (or at least he joined to maintain the project) as I understood from what he told me.
